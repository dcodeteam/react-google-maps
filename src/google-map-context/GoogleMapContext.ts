import { createContext } from "react";

export interface GoogleMapContext {
  readonly map?: google.maps.Map;
  readonly maps?: typeof google.maps;
}

// eslint-disable-next-line typescript/no-explicit-any
export const context = createContext<GoogleMapContext>({});

export const GoogleMapContextProvider = context.Provider;
export const GoogleMapContextConsumer = context.Consumer;
