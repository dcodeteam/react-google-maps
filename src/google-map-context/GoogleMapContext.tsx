import * as React from "react";

export interface GoogleMapContext {
  readonly map: google.maps.Map;
  readonly maps: typeof google.maps;
}

const { Provider, Consumer } = React.createContext<Partial<GoogleMapContext>>(
  {},
);

export interface GoogleMapContextProviderProps {
  value: GoogleMapContext;
  children: React.ReactNode;
}

export function GoogleMapContextProvider({
  value,
  children,
}: GoogleMapContextProviderProps) {
  return <Provider value={value}>{children}</Provider>;
}

export interface GoogleMapContextConsumerProps {
  children: (ctx: GoogleMapContext) => React.ReactNode;
}

export function GoogleMapContextConsumer({
  children,
}: GoogleMapContextConsumerProps) {
  return (
    <Consumer>
      {ctx =>
        !ctx.map || !ctx.maps ? null : children(ctx as GoogleMapContext)
      }
    </Consumer>
  );
}
