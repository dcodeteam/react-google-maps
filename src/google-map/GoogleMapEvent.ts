import { GenericEvent } from "../internal/GenericEvent";

export class GoogleMapEvent {
  public static readonly onClick = GenericEvent.onClick;

  public static readonly onDoubleClick = GenericEvent.onDoubleClick;

  public static readonly onRightClick = GenericEvent.onRightClick;

  public static readonly onMouseOut = GenericEvent.onMouseOut;

  public static readonly onMouseOver = GenericEvent.onMouseOver;

  public static readonly onMouseMove = GenericEvent.onMouseMove;

  public static readonly onDrag = GenericEvent.onDrag;

  public static readonly onDragStart = GenericEvent.onDragStart;

  public static readonly onDragEnd = GenericEvent.onDragEnd;

  public static readonly onIdle = "idle";

  public static readonly onTilesLoaded = "tilesloaded";

  public static readonly onTiltChanged = "tilt_changed";

  public static readonly onZoomChanged = "zoom_changed";

  public static readonly onBoundsChanged = "bounds_changed";

  public static readonly onCenterChanged = "center_changed";

  public static readonly onHeadingChanged = "heading_changed";

  public static readonly onMapTypeIdChanged = "maptypeid_changed";

  public static readonly onProjectionChanged = "projection_changed";
}
