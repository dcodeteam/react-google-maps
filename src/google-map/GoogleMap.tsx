import * as React from "react";

import { MapComponentHandlers } from "../google-map-component/MapComponentHandlers";
import {
  GoogleMapContext,
  GoogleMapContextProvider,
} from "../google-map-context/GoogleMapContext";
import { pickChangedProps } from "../internal/PropsUtils";
import { GoogleMapEvent } from "./GoogleMapEvent";

interface EventProps {
  /**
   * This handler is called when the user clicks on the map.
   *
   * An `MouseEvent` with properties for the clicked location is returned unless a place icon was clicked, in which case an `IconMouseEvent` with a `placeid` is returned.
   */
  onClick?: (
    event: google.maps.MouseEvent | google.maps.IconMouseEvent,
  ) => void;

  /**
   * This handler is called when the user double-clicks on the map.
   *
   * Note that the `onClick` handler will be also called,
   * right before this one.
   */
  onDoubleClick?: (
    event: google.maps.MouseEvent | google.maps.IconMouseEvent,
  ) => void;

  /**
   * This handler is called when the DOM context menu event is
   * fired on the map container.
   */
  onRightClick?: (
    event: google.maps.MouseEvent | google.maps.IconMouseEvent,
  ) => void;

  /**
   * This handler is called whenever the user's mouse moves over
   * the map container.
   */
  onMouseMove?: () => void;

  /**
   * This handler is called when the user's mouse exits the map container.
   */
  onMouseOut?: () => void;

  /**
   * This handler is called when the user's mouse enters the map container.
   */
  onMouseOver?: () => void;

  /**
   * This handler is repeatedly called while the user drags the map.
   */
  onDrag?: () => void;

  /**
   * This handler is called when the user stops dragging the map.
   */
  onDragEnd?: () => void;

  /**
   * This handler is called when the user starts dragging the map.
   */
  onDragStart?: () => void;

  /**
   * This handler is called when the map becomes idle after panning or zooming.
   */
  onIdle?: () => void;

  /**
   * This handler is called when the visible tiles have finished loading.
   */
  onTilesLoaded?: () => void;

  /**
   * This handler is called when the map `tilt` property changes.
   */
  onTiltChanged?: () => void;

  /**
   * This event is fired when the map `zoom` property changes.
   */
  onZoomChanged?: (event: { zoom: number }) => void;

  /**
   * This handler is called when the viewport bounds have changed.
   */
  onBoundsChanged?: (
    event: { bounds: null | undefined | google.maps.LatLngBoundsLiteral },
  ) => void;

  /**
   * This handler is called when the map `center` property changes.
   */
  onCenterChanged?: () => void;

  /**
   * This handler is called when the map `heading` property changes.
   */
  onHeadingChanged?: () => void;

  /**
   * This handler is called when the `mapTypeId` property changes.
   */
  onMapTypeIdChanged?: () => void;

  /**
   * This handler is called when the `projection` has changed.
   */
  onProjectionChanged?: () => void;
}

export interface GoogleMapProps extends EventProps {
  /**
   * Loaded `google.Maps` instance.
   */
  maps: typeof google.maps;

  /**
   * The initial Map `center`.
   */
  center: google.maps.LatLngLiteral;

  /**
   * The initial Map `zoom` level.
   */
  zoom: number;

  /**
   * The initial Map `mapTypeId`.
   */
  mapTypeId?: "HYBRID" | "ROADMAP" | "SATELLITE" | "TERRAIN";

  /**
   * Google Maps child components.
   */
  children?: React.ReactNode;

  /**
   * When `false`, map icons are not clickable.
   *
   * A map icon represents a point of interest, also known as a POI.
   */
  clickableIcons?: boolean;

  /**
   * Enables/disables `zoom` and `center` on double click.
   */
  disableDoubleClickZoom?: boolean;

  //
  // Styling
  //

  /**
   * Styles of map div.
   */
  style?: React.CSSProperties;

  /**
   * Classes of map div.
   */
  className?: string;

  /**
   * Color used for the background of the Map.
   *
   * This color will be visible when tiles have not yet
   * loaded as the user pans.
   */
  backgroundColor?: string;
}

interface State {
  ctx?: GoogleMapContext;
}

function createOptions({
  maps,

  zoom,
  center,
  backgroundColor,

  mapTypeId = "ROADMAP",
  clickableIcons = true,
  disableDoubleClickZoom = false,
}: GoogleMapProps): google.maps.MapOptions {
  return {
    disableDefaultUI: true,

    zoom,
    center,
    clickableIcons,
    backgroundColor,
    disableDoubleClickZoom,
    mapTypeId: mapTypeId && maps.MapTypeId[mapTypeId],
  };
}

function createHandlers({
  onClick,
  onDoubleClick,
  onRightClick,
  onMouseOut,
  onMouseOver,
  onMouseMove,
  onDrag,
  onDragStart,
  onDragEnd,
  onIdle,
  onTilesLoaded,
  onTiltChanged,
  onCenterChanged,
  onHeadingChanged,
  onMapTypeIdChanged,
  onProjectionChanged,
}: GoogleMapProps) {
  return {
    [GoogleMapEvent.onClick]: onClick,
    [GoogleMapEvent.onDoubleClick]: onDoubleClick,
    [GoogleMapEvent.onRightClick]: onRightClick,
    [GoogleMapEvent.onMouseOut]: onMouseOut,
    [GoogleMapEvent.onMouseOver]: onMouseOver,
    [GoogleMapEvent.onMouseMove]: onMouseMove,
    [GoogleMapEvent.onDrag]: onDrag,
    [GoogleMapEvent.onDragStart]: onDragStart,
    [GoogleMapEvent.onDragEnd]: onDragEnd,
    [GoogleMapEvent.onIdle]: onIdle,
    [GoogleMapEvent.onTilesLoaded]: onTilesLoaded,
    [GoogleMapEvent.onTiltChanged]: onTiltChanged,
    [GoogleMapEvent.onCenterChanged]: onCenterChanged,
    [GoogleMapEvent.onHeadingChanged]: onHeadingChanged,
    [GoogleMapEvent.onMapTypeIdChanged]: onMapTypeIdChanged,
    [GoogleMapEvent.onProjectionChanged]: onProjectionChanged,
  };
}

export class GoogleMap extends React.Component<GoogleMapProps, State> {
  public state: State = {};

  private mapRef = React.createRef<HTMLDivElement>();

  private handleZoomChanged = () => {
    const { ctx } = this.state;
    const { onZoomChanged } = this.props;

    if (onZoomChanged) {
      onZoomChanged({ zoom: ctx!.map.getZoom() });
    }
  };

  private handleBoundsChanged = () => {
    const { ctx } = this.state;
    const { onBoundsChanged } = this.props;

    if (onBoundsChanged) {
      const bounds = ctx!.map.getBounds();

      onBoundsChanged({ bounds: bounds && bounds.toJSON() });
    }
  };

  public componentDidMount() {
    const { maps } = this.props;
    const map = new maps.Map(this.mapRef.current);

    map.setOptions(createOptions(this.props));
    map.addListener(GoogleMapEvent.onZoomChanged, this.handleZoomChanged);
    map.addListener(GoogleMapEvent.onBoundsChanged, this.handleBoundsChanged);

    this.setState({ ctx: { map, maps } });
  }

  public componentDidUpdate(prevProps: Readonly<GoogleMapProps>): void {
    const { ctx } = this.state;

    const prevOptions = createOptions(prevProps);
    const nextOptions = createOptions(this.props);
    const options = pickChangedProps(prevOptions, nextOptions);

    if (options) {
      ctx!.map.setOptions(options);
    }
  }

  public componentWillUnmount() {
    const { ctx } = this.state;
    const {
      maps: { event },
    } = this.props;

    event.clearInstanceListeners(ctx!.map);
    event.clearInstanceListeners(this.mapRef.current!);
  }

  public render() {
    const { ctx } = this.state;
    const { style, className, children } = this.props;

    return (
      <>
        <div style={style} className={className} ref={this.mapRef} />

        {ctx != null && (
          <>
            <GoogleMapContextProvider value={ctx}>
              {children}
            </GoogleMapContextProvider>

            <MapComponentHandlers
              maps={ctx.maps}
              instance={ctx.map}
              handlers={createHandlers(this.props)}
            />
          </>
        )}
      </>
    );
  }
}
