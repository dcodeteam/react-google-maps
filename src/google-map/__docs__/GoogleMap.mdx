---
name: Map
---

import { Playground, PropsTable } from 'docz'
import { GoogleMap } from '../GoogleMap.tsx'

# Map

Draws `google.maps.Map`.

**Google Maps Docs:**
* [google.maps.Map](https://developers.google.com/maps/documentation/javascript/reference#Map)
* [google.maps.MapOptions](https://developers.google.com/maps/documentation/javascript/reference#MapOptions)

<Playground>
  <GoogleMap
    zoom={9}
    maps={google.maps}
    mapTypeId="ROADMAP"
    style={{ height: '320px' }}
    clickableIcons={true}
    disableDoubleClickZoom={false}
    center={{ lat: 36.964, lng: -122.015 }}
    onClick={event => {
      console.log('onClick', event)
    }}
    onDoubleClick={event => {
      console.log('onDoubleClick', event)
    }}
    onRightClick={event => {
      console.log('onDoubleClick', event)
    }}
  />
</Playground>

<PropsTable of={GoogleMap} />
