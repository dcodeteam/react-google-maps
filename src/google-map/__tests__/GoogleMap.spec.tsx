import { mount } from "enzyme";
import * as React from "react";

import {
  createMockHandlers,
  emitEvent,
  forEachEvent,
  getClassMockInstance,
} from "../../__tests__/testUtils";
import { GoogleMap, GoogleMapProps } from "../GoogleMap";
import { GoogleMapContextConsumer } from "../GoogleMapContext";
import { GoogleMapEvent } from "../GoogleMapEvent";

export function getMapMockInstance(): google.maps.Map {
  return getClassMockInstance(google.maps.Map);
}

describe("GoogleMap", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should attach map to child div", () => {
    const wrapper = mount(
      <GoogleMap maps={google.maps} zoom={0} center={{ lat: 0, lng: 1 }} />,
    );

    const mapDiv = wrapper.find("div");

    expect(mapDiv.length).toBe(1);
  });

  it("should pass default options to map", () => {
    mount(
      <GoogleMap maps={google.maps} zoom={0} center={{ lat: 0, lng: 1 }} />,
    );

    const map = getMapMockInstance();

    expect(map.setValues).toBeCalledTimes(1);
    expect(map.setValues).lastCalledWith({
      backgroundColor: undefined,
      center: { lat: 0, lng: 1 },
      clickableIcons: true,
      disableDefaultUI: true,
      disableDoubleClickZoom: false,
      mapTypeId: "ROADMAP",
      zoom: 0,
    });
  });

  it("should pass custom options to map", () => {
    mount(
      <GoogleMap
        maps={google.maps}
        zoom={0}
        center={{ lat: 0, lng: 1 }}
        mapTypeId="HYBRID"
        clickableIcons={false}
        disableDoubleClickZoom={true}
      />,
    );

    const map = getMapMockInstance();

    expect(map.setValues).toBeCalledTimes(1);
    expect(map.setValues).lastCalledWith({
      backgroundColor: undefined,
      center: { lat: 0, lng: 1 },
      clickableIcons: false,
      disableDefaultUI: true,
      disableDoubleClickZoom: true,
      mapTypeId: "HYBRID",
      zoom: 0,
    });
  });

  it("should add all event listeners", () => {
    mount(
      <GoogleMap zoom={0} maps={google.maps} center={{ lat: 0, lng: 1 }} />,
    );

    const map = getMapMockInstance();
    const instanceEvents = Object.keys(GoogleMapEvent).length;

    expect(map.addListener).toBeCalledTimes(instanceEvents);
  });

  it("should add listeners with handlers", () => {
    const handlers = createMockHandlers<GoogleMapProps>(GoogleMapEvent);
    const zoom = 10;

    mount(
      <GoogleMap
        {...handlers}
        zoom={zoom}
        maps={google.maps}
        center={{ lat: 0, lng: 1 }}
      />,
    );

    const map = getMapMockInstance();

    forEachEvent<GoogleMapProps>(GoogleMapEvent, (key, event) => {
      const handler = handlers[key];
      const payload = { key, event };

      expect(handler).toBeCalledTimes(0);

      emitEvent(map, event, payload);

      expect(handler).toBeCalledTimes(1);

      if (event === GoogleMapEvent.onZoomChanged) {
        expect(handler).lastCalledWith({ zoom });
      } else if (event === GoogleMapEvent.onBoundsChanged) {
        expect(handler).lastCalledWith({ bounds: map.getBounds()!.toJSON() });
      } else {
        expect(handler).lastCalledWith(payload);
      }
    });
  });

  it("pass only changed options to map", () => {
    const wrapper = mount(
      <GoogleMap maps={google.maps} zoom={0} center={{ lat: 0, lng: 1 }} />,
    );

    const map = getMapMockInstance();

    expect(map.setOptions).toBeCalledTimes(1);

    wrapper.setProps({ zoom: 1 });

    expect(map.setOptions).toBeCalledTimes(2);
    expect(map.setOptions).lastCalledWith({ zoom: 1 });

    wrapper.setProps({ zoom: 1 });

    expect(map.setOptions).toBeCalledTimes(2);

    wrapper.setProps({ zoom: 2 });

    expect(map.setOptions).toBeCalledTimes(3);

    expect(map.setOptions).lastCalledWith({ zoom: 2 });
  });

  it("should remove all listeners on unmount", () => {
    const wrapper = mount(
      <GoogleMap maps={google.maps} zoom={0} center={{ lat: 0, lng: 1 }} />,
    );

    const map = getMapMockInstance();

    const mapDiv = wrapper.find("div");

    expect(google.maps.event.clearInstanceListeners).toBeCalledTimes(0);

    wrapper.unmount();

    expect(google.maps.event.clearInstanceListeners).toBeCalledTimes(3);
    expect(google.maps.event.clearInstanceListeners).nthCalledWith(1, map);
    expect(google.maps.event.clearInstanceListeners).nthCalledWith(
      2,
      mapDiv.getDOMNode(),
    );
    expect(google.maps.event.clearInstanceListeners).nthCalledWith(3, map);
  });

  it("should pass context", () => {
    const consumer = jest.fn();

    mount(
      <GoogleMap maps={google.maps} zoom={0} center={{ lat: 0, lng: 1 }}>
        <GoogleMapContextConsumer>{consumer}</GoogleMapContextConsumer>
      </GoogleMap>,
    );

    const map = getMapMockInstance();

    expect(consumer).toBeCalledTimes(1);
    expect(consumer).toBeCalledWith({ maps: google.maps, map });
  });
});
