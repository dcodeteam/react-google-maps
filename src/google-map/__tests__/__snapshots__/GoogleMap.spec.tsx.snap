// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoogleMap #componentDidMount should add all event listeners 1`] = `
Object {
  "isThrow": false,
  "value": Map {
    "addListener": [MockFunction] {
      "calls": Array [
        Array [
          "click",
          [Function],
        ],
        Array [
          "dblclick",
          [Function],
        ],
        Array [
          "rightclick",
          [Function],
        ],
        Array [
          "mouseout",
          [Function],
        ],
        Array [
          "mouseover",
          [Function],
        ],
        Array [
          "mousemove",
          [Function],
        ],
        Array [
          "drag",
          [Function],
        ],
        Array [
          "dragstart",
          [Function],
        ],
        Array [
          "dragend",
          [Function],
        ],
        Array [
          "idle",
          [Function],
        ],
        Array [
          "tilesloaded",
          [Function],
        ],
        Array [
          "tilt_changed",
          [Function],
        ],
        Array [
          "zoom_changed",
          [Function],
        ],
        Array [
          "bounds_changed",
          [Function],
        ],
        Array [
          "center_changed",
          [Function],
        ],
        Array [
          "heading_changed",
          [Function],
        ],
        Array [
          "maptypeid_changed",
          [Function],
        ],
        Array [
          "projection_changed",
          [Function],
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    },
    "emit": [MockFunction],
    "get": [MockFunction],
    "getBounds": [MockFunction],
    "getZoom": [MockFunction],
    "listeners": Object {
      "bounds_changed": Array [
        [Function],
      ],
      "center_changed": Array [
        [Function],
      ],
      "click": Array [
        [Function],
      ],
      "dblclick": Array [
        [Function],
      ],
      "drag": Array [
        [Function],
      ],
      "dragend": Array [
        [Function],
      ],
      "dragstart": Array [
        [Function],
      ],
      "heading_changed": Array [
        [Function],
      ],
      "idle": Array [
        [Function],
      ],
      "maptypeid_changed": Array [
        [Function],
      ],
      "mousemove": Array [
        [Function],
      ],
      "mouseout": Array [
        [Function],
      ],
      "mouseover": Array [
        [Function],
      ],
      "projection_changed": Array [
        [Function],
      ],
      "rightclick": Array [
        [Function],
      ],
      "tilesloaded": Array [
        [Function],
      ],
      "tilt_changed": Array [
        [Function],
      ],
      "zoom_changed": Array [
        [Function],
      ],
    },
    "node": <div
      style="height: 100%;"
    />,
    "set": [MockFunction],
    "setValues": [MockFunction],
    "values": Object {
      "backgroundColor": undefined,
      "bounds": Array [],
      "center": Object {
        "lat": 0,
        "lng": 1,
      },
      "clickableIcons": undefined,
      "disableDefaultUI": true,
      "disableDoubleClickZoom": undefined,
      "mapTypeId": undefined,
      "zoom": 0,
    },
  },
}
`;

exports[`GoogleMap #componentDidMount should pass custom options to map: calls 1`] = `
Array [
  Array [
    <div
      style="height: 100%;"
    />,
    Object {
      "backgroundColor": undefined,
      "center": Object {
        "lat": 0,
        "lng": 1,
      },
      "clickableIcons": false,
      "disableDefaultUI": true,
      "disableDoubleClickZoom": true,
      "mapTypeId": "HYBRID",
      "zoom": 0,
    },
  ],
]
`;

exports[`GoogleMap #componentDidMount should pass default options to map: calls 1`] = `
Array [
  Array [
    <div
      style="height: 100%;"
    />,
    Object {
      "backgroundColor": undefined,
      "center": Object {
        "lat": 0,
        "lng": 1,
      },
      "clickableIcons": undefined,
      "disableDefaultUI": true,
      "disableDoubleClickZoom": undefined,
      "mapTypeId": undefined,
      "zoom": 0,
    },
  ],
]
`;
